mitodl/hugo-course-publisher
# unified content structure

#### Pre-Flight checklist


#### What are the relevant tickets?
Related to https://trello.com/c/bVQwTyeU/170-download-course-material-should-be-a-button-in-course-info

#### What's this PR do?
Firstly, this PR adds updated course markdown based on the output of [this](https://github.com/mitodl/ocw-to-hugo/pull/55) PR.  It adds the `type` and `layout` front matter keys to every page so that they are properly identified by Hugo.  The main purpose of this PR however is to remove dependency on finding the course home page through static string building.  This allows us to use the `contentDir` argument to specify a specific course to render as a single site.  I've accomplished this by creating 2 themes for the site: `multi_course` (the default) and `single_course`.  The way we've structured our content, the course info metadata lives in the markdown of the course home page, which is why we need to obtain the course home page to render content that uses this information.  Thus, inside the two theme folders you will see a `partials` folder each containing a `chp_partial.html` file.  The purpose of these partials is essentially to call another partial with the context of the course home page, which needs to be acquired differently depending on how the site is being rendered.  It's used like this:

`{{ partial "chp_partial.html" (dict "partial" "course_info.html" "context" .) }}`

By default, the site is rendered in multi-course mode.  The content in the repo is designed to be run this way.  With these changes though, you can now render a site with the root being a single course by using the Hugo `contentDir` argument.  Details about this are below.

#### How should this be manually tested?
First of all, make sure the site will still render with the normal method either using docker or node natively.  By default, it should look as if nothing has changed.  However, you can now either run the dev server or a build against a single course.  For example:

dev server:
`npm run start:single_course -- ../site/content/courses/18-06-linear-algebra-spring-2010`

build and output to `dist`:
`npm run build:hugo:single_course -- ../site/content/courses/18-06-linear-algebra-spring-2010`

Note that the path is prefixed with `..` because this is being run from the context of the node source, which is under `src`.  Either way, the result of the first command is that you should be able to run it and then visit http://localhost:3000/ and be greeted with the course home page for linear algebra.  Running the second command will simply run the Hugo build against the given course.  Verify that you can navigate the single course page just as you would if you selected it through the course index.
